### PR TITLE
Stabilize the `Saturating` type

### LINKED ISSUE #87920: Tracking Issue for `Saturating` type
This is a tracking issue for the RFC "928" (rust-lang/rfcs#928).
The feature gate for the issue is `#![feature(saturating_int_impl)]`.

### About tracking issues

Tracks progress for `Saturating` type analogue to #32463

### Unresolved Questions

 - [ ] What is the MVP for stabilization?
 - [ ] What is the correct impl of `Shl` and `ShalAssign` for `Saturating<signed>`
       https://github.com/rust-lang/rust/pull/87921#discussion_r695870065
     Possible solution: https://github.com/rust-lang/rust/issues/87920#issuecomment-913015985
     @avitex https://github.com/avitex/rfcs/blob/master/text/0000-saturating-bit-shifts.md

### Steps

- [x] implement (#87921)
- [x] Stabilize `saturating_div` (#89381)
- [ ] Split of saturating bit shifts (#103441)
- [ ] Stabilize `saturating_int_impl`

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Implementation history

 - Implementation #87921
   - Adds `struct Saturating` and `feature(saturating_int_impl)`
 - Stabilization (TODO)

### LINKED ISSUE #92354: Tracking Issue for `saturating_int_assign_impl`
Feature gate: `#![feature(saturating_int_assign_impl)]`

This is a tracking issue for adding basic integer operations to the `Saturating` Wrapping type:
```rust
let mut value = Saturating(2u8);
value += 3u8;
value -= 1u8;
value *= 2u8;
value /= 2u8;
value %= 2u8;
value ^= 255u8;
value |= 123u8;
value &= 2u8;
```

<details>
  <summary>
    <s>As well as</s>
  </summary>

  See https://github.com/rust-lang/rust/pull/93208#issuecomment-1022564429. **No longer** within the scope of this feature:

  ```rust
  let mut value = Saturating(2u8);
  let _: Saturating<u8> = value + 3u8;
  let _: Saturating<u8> = value - 1u8;
  let _: Saturating<u8> = value * 2u8;
  let _: Saturating<u8> = value / 2u8;
  let _: Saturating<u8> = value % 2u8;
  let _: Saturating<u8> = value ^ 255u8;
  let _: Saturating<u8> = value | 123u8;
  let _: Saturating<u8> = value & 2u8;
  ```
</details>

### Steps / History

 - [x] Implement basic `Saturating` type #87920
 - [x] Proposal for this implementation:  #91586
 - [x] Implementation of `saturating_int_assign_impl` #92356
   - [x]  Reduce `saturating_int_assign_impl` #93353 
 - [ ] Final comment period (FCP)
 - [ ] Stabilization PR

### Unresolved Questions

- Is this reasonable?

An implementation for the `Wrapping` type is planned as follow up.

### PR DESCRIPTION
**Pull request by kellerkindt**
Closes #87920
Closes #92354

Stabilization report https://github.com/rust-lang/rust/issues/87920#issuecomment-1652346124
FCP https://github.com/rust-lang/rust/issues/87920#issuecomment-1676438885

### CODE REVIEW

**Comment by Nilstrieb** [comment]
Thank you for your PR! You have a merge commit in your PR. Merge commits are not allowed in this repository, please [remove it](https://rustc-dev-guide.rust-lang.org/git.html#i-made-a-merge-commit-by-accident).

**Comment by kellerkindt** [comment]
Thanks, it should now be fixed.

**Review by dtolnay — APPROVED** [review]
Thanks!

File: library/core/src/num/saturating.rs [hunk 1]
```diff
@@ -24,189 +24,193 @@ use crate::ops::{Shl, ShlAssign, Shr, ShrAssign, Sub, SubAssign};
...
+//         //sh_impl_signed! { $t, i8 }
+//         //sh_impl_signed! { $t, i16 }
+//         //sh_impl_signed! { $t, i32 }
+//         //sh_impl_signed! { $t, i64 }
+//         //sh_impl_signed! { $t, i128 }
+//         //sh_impl_signed! { $t, isize }
+//     )*)
+// }
+//
+// sh_impl_all! { u8 u16 u32 u64 u128 usize i8 i16 i32 i64 i128 isize }
```
Comment (klensy) [inline library/core/src/num/saturating.rs:208]: This chunk of commented out code is expected?

**Comment by XopheD** [comment]
Is there any reason why Saturating<T> does not implement FromStr trait ?

**Comment by kellerkindt** [comment]
I don't see the point as you can do `Saturating(u16::from_str(...))`?

**Comment by XopheD** [comment]
Yes, of course, I can do it but when I use a fn which relies on FromStr, I can't do it.
```
fn f<X:FromStr>(input: &str) { ... }
```

In other words, I can’t switch the use of `u16` to `Saturating::<u16>`.

I would like to use it transparently, so stg like the following would be nice:
```
impl<X:FromStr> FromStr for Saturating<X> { ... }
```
in the same way you offer `Default` or `Display`.

But, ok, I understand that it is not so easy to decide how many standard traits you need to reimplement.

**Comment by kellerkindt** [comment]
I am not the one deciding whether this should be in the standard library or not, so I encourage you to just propose it in a separate PR to get some proper feedback :)

**Pull request merged**